Kolya Polyarniy 2537

PolarNick239 opened the pull request:
To access feeds list - press menu hardware button.

Code is bad - I know it, I have no time. May be I will do some refactoring later.

Please, next time provide task earlier - in friday, or earlier. Because we really have a lot of other tasks, especially this week, and some people really want to do android in weekend.

Some rhetorical questions about android:
- http://stackoverflow.com/questions/10302890/latest-apache-httpclient-in-android-sdk
  Android is so cool! So I can't follow your correct note at codereview - that http apache client resources should be closed(there are no EntityUtils.consume(...) and CloseableHttpResponse and so on in android now, or I do smth wrong)
- Why to access to any view object by id, we must use R.id.myButton, but not R.id.myLayout.myButton? 

Please, if my fork descriptions are negative for you or tire you - don't read them, I don't want to give you any negative emotions.

PolarNick239 committed:
RSSReader final part done.

thevery commented:
> Please, next time provide task earlier - in friday, or earlier.

Most important part of the task - database support - was clearly stated during last lesson. Other details are pretty trivial/obvious.

thevery commented:
http://developer.android.com/reference/android/net/http/AndroidHttpClient.html#close()
http://developer.android.com/reference/org/apache/http/HttpEntity.html#consumeContent()

thevery commented:
`id` is a first-class citizen on android, you can easily create id without layout at all using `/res/values/ids.xml`

PolarNick239 committed:
Now new feed's URL can be entered without "http://" prefix.
More dates formats supported - before http://news.yandex.ru/ecology.rss was not supported, now - it is supported.

PolarNick239 commented:
About id I know, what you mean, but more layouts we have - more problems we have. Why not to add (as optional posiibility) access to views id throught layouts. This is minor nuance, and I don't think that this is bad, but I think my point is useful.

http://developer.android.com/reference/org/apache/http/HttpEntity.html#consumeContent() is deprecated in actual apache client... So this is nearly a workaround.

PolarNick239 committed:
HttpEntity.consumeContent() deprecated method used.

PolarNick239 committed:
APK packaged.

PolarNick239 committed:
Usability of adding feed greatly improved! (And code improved a bit)
At first, user must enter URL, than, if feed by entered URL is not supported - user will be able to correct URL.
When URL will point to supported RSS feed, user will be asked to enter feed name(with RSS feed title as default value).

PolarNick239 committed:
Now feed name can be changed by clicking on feed in feeds managment activity.

PolarNick239 committed:
Icon added to "Manage feeds" in menu.
Some drawable resources deleted, that already contains in android.

thevery commented on RSSReader/src/main/java/com/polarnick/day06/FeedsAdapter.java:
```diff
@@ -0,0 +1,70 @@
+        ((TextView) view.findViewById(R.id.feedNameText)).setText(feed.getName());
+        ((TextView) view.findViewById(R.id.feedUrlText)).setText(feed.getUrl());
+        view.findViewById(R.id.setDefaultFeedButton).setEnabled(!feed.isDefault());
+        view.findViewById(R.id.setDefaultFeedButton).setOnClickListener(new View.OnClickListener() {
```
multiple inner classes is bad practice

PolarNick239 replied:
You mean multiple anonymous classes? I agree, I will create one inner class.

thevery commented on RSSReader/src/main/java/com/polarnick/day06/FeedsSQLiteOpenHelper.java:
```diff
@@ -0,0 +1,179 @@
+        return changedFeed;
+    }
+
+    public synchronized Map<String, FeedEntry> getEntries() {
```
why not get per feed id?

PolarNick239 replied:
Because feeds are unique by name.

thevery replied:
id is unique as well.
the point is that you should get entries for selected feed only, not all feeds.

PolarNick239 replied:
My feeds have no id.
Oh, I understood, you think that I store entries. The reason of this misunderstood is in my bad choise of names.
I store in DB only feeds' names, URLs and who is default(to be choosen on start up).

thevery replied:
no id is bad but doesn't change my point - getEntries should return list on entries, not map channel<->entries

PolarNick239 replied:
Why no id is bad? It is not normal, but in fact I don't think that this case is a case for DB using. And, my entities have their nature identificator - name.

Get entries is a cache in fact. If you mean it as public getter method - I am agree with your point.

thevery replied:
because integer (long) id is faster than text, integrates much better with cursor adapter etc
moreover, name might be not unique

PolarNick239 replied:
In my logic name must be unique - because if it will not be unique - user could not choise it from menu(he will not be able to find any difference)

I name such reasons not a bad code, but "not optimal code". :)
Seriously, if we care about peromance so much - we should not use DB for such purpose.

thevery replied:
breaking common patterns with no reason is bad practice, not suboptimal.

PolarNick239 replied:
If you mean absence of id - I agree with you, just when I was writing, I was not thinking about it as SQL table, so forget about it, and now I have not got enough time and motivation to fix this(to fix it good and with testing).

If you mean using DB everywhere - I disagree. Also your point correct not always - I mean any patterns, that became anti-patterns, singleton for example.

PolarNick239 committed:
Multiple anonymous classes fixed(setFeedDefault and deleteFeed buttons listeners).

thevery commented on RSSReader/src/main/java/com/polarnick/day06/FeedsAdapter.java:
```diff
@@ -0,0 +1,87 @@
+        ((TextView) view.findViewById(R.id.feedUrlText)).setText(feed.getUrl());
+        view.findViewById(R.id.setDefaultFeedButton).setEnabled(!feed.isDefault());
+        view.findViewById(R.id.setDefaultFeedButton).setOnClickListener(new SetDefaultButtonlistener(feed));
+        view.findViewById(R.id.deleteFeedButton).setOnClickListener(new DeleteFeedButtonListener(feed));
```
this code is bad anyway - you still create new listener for every feed

PolarNick239 replied:
I believe that this is naturally, so I don't know, why this is a bad code, and anyway I can't see any better approach. Can you suggest any? (And explain your point, if you have time and desire - I mean something other than perfomance - because perfomance is not such a criterion for code corpulence)

thevery replied:
store current feed in setTag() and make your adapter implement click listener (I agree that first part is not obvious, but second does).
new objects are not only about CPU, but - even more important - about GC.

PolarNick239 replied:
Thanks.
I know that GC will be more important in this case, I mean it under "perfomance".
I think that second idea is not obvious - because it is nearly hack IMHO. You want to make adapter more complex, I see no reasons for this, excluding VERY small perfomance improvement(one object VS two objects).
But this is IMHO - as for me, adapter, that implemens click listener looks not as good as inner listener implementation.

Moreover I think, that my point is supported by fact, that there are two buttons... But it is still possible to implement   second listener by FeedActivity for example!

Also I consider setTag as not the best solution(I mean abstract, not in android) and as a compromise for the sake of perfomance.

PolarNick239 committed:
Clicks on multiple buttons in feed managment activity now handled by one object - improved perfomance.
Minor refactoring(method renamed and new public method in FeedsSQLiteOpenHelper)

thevery commented:
menu doesn't exist on nexus 5
offline doesn't work
0
